### PR TITLE
Halting inflation

### PR DESCRIPTION
**Pull request by vzotova**
**Type of PR:**
- [X] Feature

**Required reviews:** 
- [X] 3

**What this does:**
`StakingEscrow` update only. Removes most of the code from `StakingEscrow`: halting inflation, removing all methods to control stake (including withdraw), exclude `PolicyManager` and `Adjudicator` from lifecycle (for now)

**Why it's needed:**
First step in merging networks

**Notes for reviewers:**
Only tests for contracts have been adjusted. Some code for `PolicyManger` and `Adjudicator` was not removed to use it together with `PREApplication` contract later

### CODE REVIEW

File: nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol [hunk 1]
```diff
@@ -109,9 +67,9 @@ contract StakingEscrowStub is Upgradeable {
 * @title StakingEscrow
 * @notice Contract holds and locks stakers tokens.
 * Each staker that locks their tokens will receive some compensation
-* @dev |v5.7.1|
+* @dev |v6.1.1|
```
[the remark this answers is not in the document]
Comment (vzotova) [reply nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol:70]: I see that we are using min version as 1 (for previous contracts). But honestly can't remember 😅

File: nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol [hunk 2]
```diff
@@ -203,45 +101,13 @@ contract StakingEscrow is Issuer, IERC900History {
...
     * @param snapshotsEnabled Updated parameter value
     */
     event SnapshotSet(address indexed staker, bool snapshotsEnabled);
 
-    /**
-    * @notice Signals that the staker migrated their stake to the new period length
-    * @param staker Staker address
-    * @param period Period when migration happened
-    */
-    event Migrated(address indexed staker, uint16 indexed period);
```
[the remark this answers is not in the document]
Comment (vzotova) [reply nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol:240]: We will not have split time in periods anymore so no difference between migrated and non-migrated stakes

[the remark this answers is not in the document]
Comment (derekpierre) [reply nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol:240]: 👍

File: nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol [hunk 3]
```diff
@@ -425,249 +215,32 @@ contract StakingEscrow is Issuer, IERC900History {
...
-    function getWorkerFromStaker(address _staker) external view returns (address) {
-        return stakerInfo[_staker].worker;
     }
 
     /**
     * @notice Get work that completed by the staker
     */
     function getCompletedWork(address _staker) external view returns (uint256) {
-        return stakerInfo[_staker].completedWork;
+        return token.totalSupply();
```
Comment (cygnusv) [inline nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol:229]: Just to double-check: this implies that we consider all pending work in WorkLock to be done, so people that still have ETH locked can claim it back. Correct? In such case, I think this should be included in the newsfragment.

File: nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol [hunk 4]
```diff
@@ -681,57 +254,19 @@ contract StakingEscrow is Issuer, IERC900History {
...
-                continue;
-            }
-
-            subStake.unlockingDuration = _windDown ? subStake.unlockingDuration - 1 : subStake.unlockingDuration + 1;
-            if (subStake.unlockingDuration == 0) {
-                subStake.lastPeriod = nextPeriod;
-            }
-        }
+        addSnapshot(info, int256(_value));
+        emit Deposited(_staker, _value);
```
Comment (cygnusv) [inline nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol:269]: So, we are removing the possibility of depositing NU to StakingEscrow, but we're still allowing WorkLock deposits. Can this have implications on the Threshold side? Maybe a possibility is to somehow end WL completely (i.e. refund & no more deposits)?

Comment (vzotova) [reply nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol:269]: there is no any way to stop `WorkLock`, so I can't see how to avoid deposit. Re Threshold: all these "new" tokens are in `WorkLock` contract and they should be part of `currentPeriodSupply`. For me it looks like we already consider them as deposited in `StakingEscrow`

File: nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol (hunk 3 above)
Comment (vzotova) [reply nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol:229]: ✔️

File: nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol (hunk 4 above)
Comment (cygnusv) [reply nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol:269]: > all these "new" tokens are in WorkLock contract and they should be part of currentPeriodSupply. For me it looks like we already consider them as deposited in StakingEscrow

OK, this is a good reasoning.

**Review by cygnusv — APPROVED** [review]
A PR for the history books. Thanks @vzotova!

File: newsfragments/2822.feature.rst [hunk 5]
```diff
@@ -0,0 +1 @@
+Halting NU inflation, now refund in WorkLock is possible without work (claim stil needed)
```
Comment (derekpierre) [inline newsfragments/2822.feature.rst:1]:
```suggestion
Halting NU inflation; now refund in WorkLock is possible without work (claim still needed)
```

**Review by derekpierre — COMMENTED** [review]
@vzotova - I assume the failing tests will be dealt with separately as part of the overall EPIC...?

File: newsfragments/2822.feature.rst (hunk 5 above)
Comment (vzotova) [reply newsfragments/2822.feature.rst]: ✔️

**Comment by vzotova** [comment]
> @vzotova - I assume the failing tests will be dealt with separately as part of the overall EPIC...?

Yeah, it's hard to fix those tests before PREApp contract will be ready (at least)

File: nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol (hunk 4 above)
Comment (arjunhassard) [reply nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol:269]: No bids after Sep 28th 2020 (https://dune.xyz/queries/271579) and only one deposit on Aug 30th 2020 (https://dune.xyz/queries/271593). Would the only implication be that someone loses money for (inexplicably) sending in ETH and there's no longer away to call `worktoETH`?

File: tests/contracts/contracts/StakingContractsTestSet.sol [hunk 6]
```diff
@@ -287,3 +234,41 @@ contract SimpleStakingContract is AbstractStakingContract, Ownable {
...
+        NuCypherToken _token,
+        IStakingEscrow _escrow,
+        PolicyManager _policyManager,
+        WorkLock _workLock
+    )
+        StakingInterface(_token, _escrow, _policyManager, _workLock)
+    {
+    }
+
+    function depositAsStaker(uint256 _value, uint16 _periods) public onlyDelegateCall {
```
Comment (arjunhassard) [inline tests/contracts/contracts/StakingContractsTestSet.sol:262]: Are these tests for deposit/withdraw methods that will be used post-merge in some way? How is this related to the staking adaptor contract set?

**Review by arjunhassard — APPROVED** [review]
End of an era 🥲

File: nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol (hunk 4 above)
Comment (vzotova) [reply nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol:269]: I don't understand the question. Both events is only for initial worklock phase which was more than year ago

File: tests/contracts/contracts/StakingContractsTestSet.sol (hunk 6 above)
Comment (vzotova) [reply tests/contracts/contracts/StakingContractsTestSet.sol:262]: These lines to minimize changes in tests right now

File: nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol (hunk 4 above)
Comment (arjunhassard) [reply nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol:269]: Yep, that's what I was getting at with my question – what exactly is the undesirable corner case where someone makes a deposit post initial worklock phase? Has that occurred in the last 13 months?

Comment (vzotova) [reply nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol:269]: This deposit for WorkLock, not for StakingEscrow
